dali/python/nvidia/dali/pool.py

```diff
@@ -0,0 +1,457 @@
+from nvidia.dali import shared_mem
+
+
+class SharedBatchesConsumer:
+    """Counterpart of worker.py:SharedBatchesDispatcher. Can receive and deserialize batch
+    from the worker, keeps track of already received memory chunks and opens new chunks
+    or resizes exiting ones if necessary."""
+    class MemChunk:
+        def __init__(self, shm_chunk: shared_mem.SharedMem, capacity: int):
+            self.shm_chunk = shm_chunk
+            self.capacity = capacity
+
+    def __init__(self):
+        import_numpy()
+        self.batch_pool = {}
+
+    def get_mem_chunk(self, sock: socket.socket, batch: SharedBatchMeta) -> MemChunk:
+        """Get the handle for shared memory through sock and mmap the memory based on metadata
+        in `batch`. Adjust if it was previously mapped.
+
+        Parameters
+        ----------
+        `sock` : socket.socket
+            Socket used to transfer shared memory handles.
+        `batch` : SharedBatchMeta
+            Serialized batch metadata
+
+        Returns
+        -------
+        MemChunk
+            Wrapped SharedMem object with current capacity containing the obtained data.
+        """
+        chunk = self.batch_pool.get(batch.mem_chunk_id)
+        if chunk is not None:
+            if chunk.capacity != batch.capacity:
+                chunk.shm_chunk.resize(batch.capacity, trunc=False)
+                chunk.capacity = batch.capacity
+            return chunk
+        handle, shm_chunk = -1, None
+        try:
+            handle = multiprocessing.reduction.recv_handle(sock)
+            # TODO(windows): We're pretending here that the handle is not a fd, which in fact
+            # it is. The call below probably needs to be adjusted.
+            assert os.fstat(handle).st_size >= batch.capacity
+            shm_chunk = shared_mem.SharedMem.open(handle, batch.capacity)
+        except:
+            if shm_chunk is not None:
+                shm_chunk.close()
+            # close handle manually if shm_chunk creation failed, otherwise shm_chunk
+            # is responsible for doing so
+            elif handle >= 0:
+                os.close(handle)
+            raise
+        chunk = self.MemChunk(shm_chunk, batch.capacity)
+        self.batch_pool[batch.mem_chunk_id] = chunk
+        return chunk
+
+    def load_batch(self, sock: socket.socket, batch_meta: SharedBatchMeta):
+        """Based on the metadata in `batch` obtain the smem mapping, obtain the sample metadata
+        and deserialize the resulting data.
+
+        Returns
+        -------
+        List of (sample id: int, sample)
+            Indexed samples from the obtained part of batch
+        """
+        chunk = self.get_mem_chunk(sock, batch_meta)
+        # Load list of indexed SampleMeta: (idx, SampleMeta) or (idx, tuple of SampleMeta)
+        sample_meta = deserialize_sample_meta(chunk.shm_chunk, batch_meta)
+        return deserialize_batch(chunk.shm_chunk, sample_meta)
```

Comment:
Once again, why we cannot fuse `deserialize_sample_meta` and `deserialize_batch` into one function? Do we need sample_meta to be accessible outside of these function?

Reply:
I guess we can.